pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -620,258 +574,264 @@ def multi_modal_content_identifier(identifier: str | bytes) -> str:
+                    tool_result_part, extra_parts = task.result()
+                    yield _messages.FunctionToolResultEvent(tool_result_part)
+
+                    parts_by_index[index] = [tool_result_part, *extra_parts]
+
+        # We append the results at the end, rather than as they are received, to retain a consistent ordering
+        # This is mostly just to simplify testing
+        for k in sorted(parts_by_index):
+            output_parts.extend(parts_by_index[k])
+
+    # Finally, we handle deferred tool calls
+    for call in tool_calls_by_kind['deferred']:
+        if final_result:
+            output_parts.append(
+                _messages.ToolReturnPart(
+                    tool_name=call.tool_name,
+                    content='Tool not executed - a final result was already processed.',
+                    tool_call_id=call.tool_call_id,
+                )
+            )
+        else:
+            yield _messages.FunctionToolCallEvent(call)
 
     output_parts.extend(user_parts)
 
+    if final_result:
+        output_final_result.append(final_result)
 
-async def _tool_from_mcp_server(
-    tool_name: str,
-    ctx: GraphRunContext[GraphAgentState, GraphAgentDeps[DepsT, NodeRunEndT]],
-) -> Tool[DepsT] | None:
-    """Call each MCP server to find the tool with the given name.
 
-    Args:
-        tool_name: The name of the tool to find.
-        ctx: The current run context.
+async def _call_function_tool(
+    toolset: AbstractToolset[DepsT],
+    tool_call: _messages.ToolCallPart,
+    run_context: RunContext[DepsT],
+    tracer: Tracer,
+    include_content: bool = False,
+) -> tuple[_messages.ToolReturnPart | _messages.RetryPromptPart, list[_messages.ModelRequestPart]]:
+    """Run the tool function asynchronously.
 
-    Returns:
-        The tool with the given name, or `None` if no tool with the given name is found.
+    See <https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/#execute-tool-span>.
     """
+    span_attributes = {
+        'gen_ai.tool.name': tool_call.tool_name,
+        # NOTE: this means `gen_ai.tool.call.id` will be included even if it was generated by pydantic-ai
+        'gen_ai.tool.call.id': tool_call.tool_call_id,
+        **({'tool_arguments': tool_call.args_as_json_str()} if include_content else {}),
+        'logfire.msg': f'running tool: {tool_call.tool_name}',
+        # add the JSON schema so these attributes are formatted nicely in Logfire
+        'logfire.json_schema': json.dumps(
+            {
+                'type': 'object',
+                'properties': {
+                    **(
+                        {
+                            'tool_arguments': {'type': 'object'},
+                            'tool_response': {'type': 'object'},
+                        }
+                        if include_content
+                        else {}
+                    ),
+                    'gen_ai.tool.name': {},
+                    'gen_ai.tool.call.id': {},
```

Comment:
I'm guessing you didn't change these, just copy-pasted, but it seems weird to me that these lines are even necessary...

Reply:
Yeah this originally came from https://github.com/pydantic/pydantic-ai/pull/1158, I don't know exactly how `logfire.json_schema` is handled but I agree we probably don't need to list properties that we don't specify a type for anyway.